pageserver/src/tenant/mgr.rs

```diff
@@ -279,24 +280,44 @@ pub async fn create_tenant(
     ctx: &RequestContext,
 ) -> Result<Arc<Tenant>, TenantMapInsertError> {
     tenant_map_insert(tenant_id, |vacant_entry| {
+        let succeed = Cell::new(false);
+
         // We're holding the tenants lock in write mode while doing local IO.
         // If this section ever becomes contentious, introduce a new `TenantState::Creating`
         // and do the work in that state.
         let tenant_directory = super::create_tenant_files(conf, tenant_conf, tenant_id, CreateTenantFilesMode::Create)?;
-        // TODO: tenant directory remains on disk if we bail out from here on.
-        //       See https://github.com/neondatabase/neon/issues/4233
+
+        scopeguard::defer! {
+            if !succeed.get() {
+                std::fs::remove_dir_all(&tenant_directory).ok(); // ignore the error
+            }
+        }
 
         let created_tenant =
             schedule_local_tenant_processing(conf, &tenant_directory, remote_storage, ctx)?;
-        // TODO: tenant object & its background loops remain, untracked in tenant map, if we fail here.
-        //      See https://github.com/neondatabase/neon/issues/4233
+
+        let tenant_cloned = Arc::clone(&created_tenant);
+        scopeguard::defer! {
+            if !succeed.get() {
+                // As we might have removed the directory, the tenant should directly go into the broken state.
+                tenant_cloned.set_broken("failed to create".into());
```

Comment:
I haven't checked, do the background tasks stop if tenant becomes broken or will they log some stacktrace continiously?

Reply:
They will log some error but only once.

Reply:
Sometimes there will be error like "directory not found" because we have already removed it, and sometimes it will be "tenant state is broken". After that they will stop.